### ISSUE TITLE
Equivalence between Patch Merging and Conv.

### ISSUE THREAD

**Issue by AHHOZP** [microsoft/Swin-Transformer#256]
Hello, after looking at the code in patch merging part, we found the complex operation that slice the feature and concatenate them then go through the linear layer to reduce dimension from 4C to 2C is completely equal to a conv layer of kernel size 2 and stride 2.

The operation you did is concatenate 4 pixels from a 2x2 patch in to 1 pixel, but quadrupled channel.
Every 2x2 patch shared the same weight with other patches in your linear layer (self.reduction).
The conv(kernel size=2, stride=2) does the same thing.

Amount of parameters of this linear layer is equal to this conv layer.
linear layer params = input channel * output channel = 4C * 2C = 8 * C^2
conv layer params = kernel size * kernel size * input channel * output channel = 2 * 2 * C * (2 * C) = 8 * C^2
SO, linear layer params == conv layer params

**Comment by yan-mingyuan** [comment]
I think both implementations are essentially the same. 
The first approach uses slicing, which might seem intricate at first, but after optimization, it adds minimal overhead.
On the other hand, the second approach involves conv2d, which looks more elegant. However, the inclusion of the permute operation could potentially introduce additional overhead.

**Comment by AHHOZP** [comment]
> 

You are right, but the tensor slice and concat also cost time. I dont know how much, and which cost more when compared to contiguous.
Another problem is layernorm, the code for 'Using Conv2d', you will find the norm is different from origin implementation.
And batchnorm is used for cnn instead of layernorm, so I guess this may be another reason why they use patch mergy instead of cnn.

**Comment by dongdongtong** [comment]
This is very interesting. I am thinking that in swinv1, the conv operation is not equal to the slice patching because swinv1 use pre norm in its merging operation, which leads to the norm not equaling to the conv patch merging.

But in swinv2, the author used post norm, which first conducted reduction and then did the norm. If we conducted the conv patch merging, this will completely equal to the slice patch merging, right?
